### PR TITLE
Graph Activity feed bulk API - C#

### PR DESCRIPTION
**Pull request by HunaidHanfee-MSFT**

### CODE REVIEW

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs [hunk 1]
```diff
@@ -110,19 +109,56 @@ public async Task<ActionResult> SendNotificationToOrganisation(BroadcastInfo tas
...
+                    catch (Exception ex)
+                    {
+                        Console.WriteLine("Installation failed for " + users.DisplayName);
+                    }
+                });
+
+                var url = "https://teams.microsoft.com/l/entity/" + appId.ToList()[0] + "/broadcast?context={\"subEntityId\":\"" + taskInfo.taskId + "\"}";
+                var client = new HttpClient();
+
+                var recipientsChunks = UtilityHelper.SplitIntoChunks(recipientsList, 85);
```
Comment (ChetanSharma-msft) [inline samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs:158]: What is 85 number ?
Can we add constant or variable for it and use?

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs [hunk 2]
```diff
@@ -110,19 +109,56 @@ public async Task<ActionResult> SendNotificationToOrganisation(BroadcastInfo tas
...
-                var appId = installedAppsForCurrentUser.Where(id => id.TeamsAppDefinition.DisplayName == "Activity feed broadcast").Select(x => x.TeamsAppDefinition.TeamsAppId);
+                var appId = installedAppsForCurrentUser.Where(id => id.TeamsAppDefinition.DisplayName == _configuration["AppName"]).Select(x => x.TeamsAppDefinition.TeamsAppId);
+                var count = usersList.Count;
+                var counter = 0;
+                var recipientsList = new List<Dictionary<string, string>>();
 
-                Parallel.ForEach (usersList, async users =>
-                {
-                    var installedApp = await graphClient.Users[users.Id].Teamwork.InstalledApps
+                await UtilityHelper.ForEachAsync(usersList, 5, async users => {
```
Comment (ChetanSharma-msft) [inline samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs:117]: What is 5 as an number?
Can we add constant or variable for it and use?

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs [hunk 3]
```diff
@@ -110,19 +109,56 @@ public async Task<ActionResult> SendNotificationToOrganisation(BroadcastInfo tas
...
+                            await graphClient.Users[users.Id].Teamwork.InstalledApps
+                                                .Request()
+                                                .AddAsync(userScopeTeamsAppInstallation);
+                        }
+                        counter++;
+                        recipientsList.Add(new Dictionary<string, string>
+                        {
+                            {"@odata.type", "microsoft.graph.aadUserNotificationRecipient" },
+                            {"userId", users.Id}
+                        });
```
Comment (ChetanSharma-msft) [inline samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs:147]: Add proper line spacing & break

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs [hunk 4]
```diff
@@ -110,19 +109,56 @@ public async Task<ActionResult> SendNotificationToOrganisation(BroadcastInfo tas
...
+                        Console.WriteLine("Installation failed for " + users.DisplayName);
+                    }
+                });
+
+                var url = "https://teams.microsoft.com/l/entity/" + appId.ToList()[0] + "/broadcast?context={\"subEntityId\":\"" + taskInfo.taskId + "\"}";
+                var client = new HttpClient();
+
+                var recipientsChunks = UtilityHelper.SplitIntoChunks(recipientsList, 85);
+
+                foreach (var recipientChunk in recipientsChunks)
```
Comment (ChetanSharma-msft) [inline samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs:160]: Rename it to recipient

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Helpers/UtilityHelper.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        /// <summary>
+        /// Process the data in parallel foreach loop.
+        /// </summary>
+        /// <typeparam name="T">Type of source data.</typeparam>
+        /// <param name="source">Source data.</param>
+        /// <param name="dop">Partition count.</param>
+        /// <param name="body">Function which process the data.</param>
+        /// <returns></returns>
+        public static Task ForEachAsync<T>(IEnumerable<T> source, int dop, Func<T, Task> body)
```
Comment (ChetanSharma-msft) [inline samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Helpers/UtilityHelper.cs:23]: What is dop ??
Do you mean top ?

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Helpers/UtilityHelper.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+        /// <summary>
+        /// Splits a <see cref="List{T}"/> into multiple chunks.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="list">The list to be chunked.</param>
+        /// <param name="chunkSize">The size of each chunk.</param>
+        /// <returns>A list of chunks.</returns>
+        public static List<List<T>> SplitIntoChunks<T>(List<T> list, int chunkSize)
```
Comment (ChetanSharma-msft) [inline samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Helpers/UtilityHelper.cs:42]: list => recipients or users

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Helpers/UtilityHelper.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            if (chunkSize <= 0)
+            {
+                throw new ArgumentException("chunkSize must be greater than 0.");
+            }
+
+            List<List<T>> retVal = new List<List<T>>();
```
Comment (ChetanSharma-msft) [inline samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Helpers/UtilityHelper.cs:49]: Users or recipients

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs (hunk 4 above)
Comment (ChetanSharma-msft) [inline samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs:160]: Suppose each chunk has 10 records & suppose you get 4 chunks.
How are you iterating for every chunk recipients as I could see only one loop ?

File: samples/graph-activity-feed-broadcast/csharp/README.md [hunk 8]
```diff
@@ -91,7 +91,7 @@ This sample shows a feature where user can broadcast a message to all members of
     Set a redirect URI:
     * Select **Add a platform**.
     * Select **web**.
-    * Enter the **redirect URI** for the app in the following format: `https://{Base_Url}/Auth/end`, `https://{Base_Url}/Auth/Start`. This will be the page where a successful implicit grant flow will redirect the user.
+    * Enter the **redirect URI** for the app in the following format: `https://{Base_Url}/Auth/End`, `https://{Base_Url}/Auth/Start`. This will be the page where a successful implicit grant flow will redirect the user.
     Enable implicit grant by checking the following boxes:  
     ✔ ID Token
```
Comment (ChetanSharma-msft) [inline samples/graph-activity-feed-broadcast/csharp/README.md:96]: Can you please add the doc link for sending bulk notification below in further reading section ?

**Review by ChetanSharma-msft — CHANGES_REQUESTED** [review]
Reviewed and provided my comments

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs (hunk 4 above)
Comment (HunaidHanfee-MSFT) [reply samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs:160]: It contains list of chunks of participant. That means recipientsChunks would look like this `[[user1, user2, user 3], [user4, user5, user 6]]`, We can assign these individuals chunk to Graph API so using one loop to iterate through recipientsChunks  and directly assign individual chunk `recipientChunk` in API.

File: samples/graph-activity-feed-broadcast/csharp/README.md (hunk 8 above)
Comment (HunaidHanfee-MSFT) [reply samples/graph-activity-feed-broadcast/csharp/README.md:96]: Added

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Helpers/UtilityHelper.cs (hunk 7 above)
Comment (HunaidHanfee-MSFT) [reply samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Helpers/UtilityHelper.cs:49]: It is a generic helper class, so renamed as `chunksList` as it represents list of chunks of element.

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Helpers/UtilityHelper.cs (hunk 6 above)
Comment (HunaidHanfee-MSFT) [reply samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Helpers/UtilityHelper.cs:42]: It is generic helper class so not marking as specific to user or recipients. Can be used for any type.

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs (hunk 4 above)
Comment (HunaidHanfee-MSFT) [reply samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs:160]: Updated to recipientList as its not individual entity.

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs (hunk 3 above)
Comment (HunaidHanfee-MSFT) [reply samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs:147]: Done.

**Review by ChetanSharma-msft — APPROVED** [review]
Changes looks fine, approving the PR.

**Review by Wajeed-msft — APPROVED** [review]

**Pull request merged**